packages/eslint-plugin/tests/docs.test.ts

```diff
@@ -46,7 +59,72 @@ function tokenIsH2(
   );
 }
 
+function renderLintResults(code: string, errors: Linter.LintMessage[]): string {
+  const output: string[] = [];
+  const lines = code.split(/\r?\n/);
+
+  for (let i = 0; i < lines.length; i++) {
+    const line = lines[i];
+    output.push(line);
+
+    for (const error of errors) {
+      const startLine = error.line - 1;
+      const endLine =
+        error.endLine === undefined ? startLine : error.endLine - 1;
+      const startColumn = error.column - 1;
+      const endColumn =
+        error.endColumn === undefined ? startColumn : error.endColumn - 1;
+      if (i < startLine || i > endLine) {
+        continue;
+      }
+      if (i === startLine) {
+        const squiggle = '~'.repeat(
+          startLine === endLine
+            ? Math.max(1, endColumn - startColumn)
+            : line.length - startColumn,
+        );
+        const squiggleWithIndent = ' '.repeat(startColumn) + squiggle + ' ';
+        const errorMessageIndent = ' '.repeat(squiggleWithIndent.length);
+        output.push(
+          squiggleWithIndent +
+            error.message.split('\n').join('\n' + errorMessageIndent),
+        );
+      } else if (i === endLine) {
+        output.push('~'.repeat(endColumn));
+      } else {
+        output.push('~'.repeat(line.length));
+      }
+    }
+  }
+
+  return output.join('\n').trim() + '\n';
+}
+
+const linter = new Linter();
+linter.defineParser('@typescript-eslint/parser', tseslintParser);
+
+const eslintOutputSnapshotFolder = path.resolve(
+  __dirname,
+  'docs-eslint-output-snapshots',
+);
+fs.mkdirSync(eslintOutputSnapshotFolder, { recursive: true });
+
 describe('Validating rule docs', () => {
+  let fromMarkdown: typeof FromMarkdown;
+  let mdxjs: typeof Mdxjs;
+  let mdxFromMarkdown: typeof MdxFromMarkdown;
+  let unistUtilVisit: typeof UnistUtilVisit;
+  beforeAll(async () => {
+    // dynamic import('...') is transpiled to the require('...') call,
+    // but all modules imported below are ESM only, so we cannot require() them
+    // eslint-disable-next-line @typescript-eslint/no-implied-eval
+    const dynamicImport = new Function('module', 'return import(module)');
+    ({ fromMarkdown } = await dynamicImport('mdast-util-from-markdown'));
+    ({ mdxjs } = await dynamicImport('micromark-extension-mdxjs'));
+    ({ mdxFromMarkdown } = await dynamicImport('mdast-util-mdx'));
+    unistUtilVisit = await dynamicImport('unist-util-visit');
+  });
```

Comment:
question: could we change this file's extension to `.mts` so that we can use ESM and avoid this hack?

Reply:
Ah, I've tried it!

But if we change it to `.mts`, we get a bunch of other issues:

- we need to add `.js` extension to relative imports because of `moduleResolution`
- we have to remove these `.js` extensions in jest via `moduleNameMapper` (jest does not load these files without `moduleNameMapper`)
- these imported files contain all exported members in the `default` property, so I added `interopDefault`

Here I tried to address all of them - https://github.com/auvred/typescript-eslint/commit/020b12cbaaf806c12cfe3d566f05c11cef36f72b (`yarn lint`, `yarn test` and `yarn typecheck` pass locally)

I'm not sure which solution is better (I don't like both, though :smile:)

Reply:
Ugh jest is becoming a burden.
Might be about time to consider migrating to something else that supports ESM better?